crates/matrix-sdk-ui/src/timeline/pagination.rs

```diff
@@ -12,9 +12,108 @@
+                let notified = pin!(self.sync_response_notify.notified());
+                match timeout(notified, Duration::from_secs(3)).await {
+                    Ok(()) => match self.inner.back_pagination_token().await {
+                        Some(token) => Some(token),
+                        None => {
+                            warn!(
+                                "Sync response without prev_batch received, \
+                                 start of timeline reached?"
+                            );
+                            return Ok(ControlFlow::Break(BackPaginationStatus::Idle));
+                        }
+                    },
+                    Err(_) => {
+                        debug!("Waiting for prev_batch token timed out after 3s");
+                        None
+                    }
+                }
+            }
+            token => token,
+        };
+
+        self.back_pagination_status.set_if_not_eq(BackPaginationStatus::Paginating);
+
+        let mut outcome = PaginationOutcome::new();
+        while let Some(limit) = options.next_event_limit(outcome) {
+            let messages = self
+                .room()
+                .messages(assign!(MessagesOptions::backward(), {
+                    from: from.clone(),
+                    limit: limit.into(),
+                }))
+                .await?;
+            let chunk_len = messages.chunk.len();
+
+            let tokens = PaginationTokens { from, to: messages.end.clone() };
+            let res = match self.inner.handle_back_paginated_events(messages.chunk, tokens).await {
+                Ok(result) => Some(result),
+                Err(HandleBackPaginatedEventsError::TokenMismatch) => {
+                    info!("Start of timeline was altered since pagination was started, resetting");
+                    return Ok(ControlFlow::Continue(()));
+                }
+                Err(HandleBackPaginatedEventsError::ResultOverflow) => None,
+            };
+
+            let process_events_result = async {
+                let res = res?;
+
+                outcome.events_received = chunk_len.try_into().ok()?;
+                outcome.total_events_received =
+                    outcome.total_events_received.checked_add(outcome.events_received)?;
+
+                outcome.items_added = res.items_added;
+                outcome.items_updated = res.items_updated;
+                outcome.total_items_added =
+                    outcome.total_items_added.checked_add(outcome.items_added)?;
+                outcome.total_items_updated =
+                    outcome.total_items_updated.checked_add(outcome.items_updated)?;
+
+                Some(())
+            }
+            .await;
+
+            from = messages.end;
+
+            if from.is_none() {
+                break;
+            }
+
+            if process_events_result.is_none() {
+                error!("Received an excessive number of events, ending pagination (u16 overflow)");
```

Comment:
nit: we could remove the "u16 overflow" part, it's an implementation detail that might not ring any bell to a consumer of this API

Reply:
Hmm, okay.